pihole: init at various

averyvigolo opened the pull request:
###### Description of changes
Recreated from #224881

[Pihole](https://pi-hole.net) describes itself as "A black hole for Internet advertisements".

It's a suite for running and administrating a DNS sinkhole, which filters queries and caches their results, primarily to block advertisements and other undesirable traffic originating from the LAN.

Built on dnsmasq, the DNS server (called pihole-FTL) can also be used as a DHCP server to automatically configure LAN clients to use pihole's DNS.

The pihole-ftl package is based on #108055 by @JamieMagee (itself based on the work of @nuxeh). The remaining work was done mostly from scratch.

This PR resolves the package request issue #61617.

This PR includes packages for:

* [pihole-ftl](https://github.com/pi-hole/FTL): The DNS server, forked from dnsmasq
* [pihole-web:](https://github.com/pi-hole/web) A web server which shows statistics, logs, and allows more advanced configuration of pihole-FTL
* [pihole](https://github.com/pi-hole/pi-hole): A collection of scripts for managing pihole-FTL

The last one needs the heaviest modifications, particularly to excise the reinstallation/upgrade functionality, which is not needed on NixOS.

NixOS modules are provided for pihole-ftl and pihole-web.

pihole-ftl's module is largely a wrapper around the existing dnsmasq one, since they are mostly compatible applications. Some additional options are provided for automatically setting the adlists for pihole to download and apply.

pihole-web's module is rather minimal, with some cosmetic options for temperature units and the web theme.

To do:

- [ ] Investigate `pihole setpassword` not working

- [ ] NixOS tests for pihole-web and pihole-FTL?

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

averyvigolo committed:
pihole-ftl: init at 6.1

Replace a hardcoded execv with a dynamic execvp to allow the executed program to
live in the Nix store.

Co-Authored-By: JamieMagee <jamie.magee@gmail.com>
Co-Authored-By: nuxeh <1516017+nuxeh@users.noreply.github.com>

averyvigolo committed:
pihole: init at 6.1

Which is the core admin script for pihole, which modifies the pihole-FTL and
gravity (DNS settings) database.

It requires quite a lot of patching to remove most of the hard-coded absolute
paths, and add the many dependencies the script has.

Patches also remove numerous unnecessary/unsupported operations from the script,
including reinstalling, uninstalling, and updating pihole, pihole-FTL, and
pihole-web.

Some scripts also include the installation scripts for some utilities, so these
utils need to be copied over to the files which are included in the Nix package.

averyvigolo committed:
pihole-web: init at 6.1

averyvigolo committed:
dnsmasq: pass through config file as an option

So that the pihole module reuse it. pihole-ftl is a fork of dnsmasq and its
configuration file is compatible.

fricklerhandwerk commented on nixos/modules/services/networking/pihole-ftl.md:
```diff
@@ -0,0 +1,54 @@
+
+*Upstream documentation*: <https://docs.pi-hole.net/>
+
+[Pi-hole](https://pi-hole.net/) is a suite for creating and managing a local DNS
```
Please break lines on sentence boundaries. Arbitrary breaks make review suggestion diffs needlessly messy.

williamvds replied:
Sure, I've fixed this, thanks

Pandapip1 commented:
I don't see the adminlte package?

williamvds commented:
> I don't see the adminlte package?

Ah, it was renamed to pihole-web after an upstream rename. Updated the description

purepani commented:
Could you fix the merge conflicts?

I'll review it once that's done

williamvds commented:
@purepani cheers! Rebased

There's been a major pihole upgrade to 6.0 in the meantime, so I'll work on packaging the update. Would we be happy to merge this older version 5?

purepani commented:
Personally, I'd say just do pihole 6.0 before merging, depending on how much time it would take.

williamvds commented:
Alrighty, will give it a quick shot

purepani reviewed (commented):
Wrote a few comments on the service; didn't look at the packages since they will probably change quite a bit :)

purepani commented on nixos/modules/services/web-apps/pihole-web.nix:
```diff
@@ -0,0 +1,144 @@
+      ];
+    };
+
+    environment.etc."pihole/dns-servers.conf" = {
+      user = ftlCfg.user;
+      group = ftlCfg.group;
+      text = cfg.dnsServers;
+    };
```
Is this location not configurable in pihole? Generally, I would say it's better to put this in a store location, and point pihole to that.

williamvds replied:
Previously, no, most of these paths are hardcoded to `/etc/pihole` and could not be changed.

With v6 things are a bit better - you can specify the paths for certain database files, but I think the daemon still puts some files in `/etc/pihole`.
`/etc/pihole/pihole.toml` is hardcoded as far as I can tell.
I'll change the default config to move what can be configured out of `/etc/pihole` into `/var/lib/pihole`, which is where database files etc belong really

williamvds commented:
Thanks for the initial comments!

I've made some progress on the updates.
Not many major changes to the Nix packages and modules, just patches and dependencies have changed a bit for the main pihole and pihole-ftl packages.

The web frontend seems to have changed from a regular PHP app to use `.lp` files, which at the least is a different framework than usual. I'll have to dig a bit more into those tomorrow.

williamvds commented:
v6 has changed the config system to a single TOML file, which is nice and centralised but does mean redoing the config in the service module.

The dhcpd fork "FTL" now also serves the web server, which is slightly horrifying, but should simplify packaging somewhat.

williamvds commented:
Pushed the current progress. I'll rebase once you've done with the reviews to hopefully make the diffs easier to read - I know GitHub doesn't deal with that well.

Some TODOs:

- [ ] Fix the Update Gravity page on the web GUI
- [ ] Fix the adlist option so users can declare adlists
  - For some reason upstream removed the handy script option I previously used
- [ ] Fix the nixpkgs docs (redirects)
- [ ] Somehow add the full autogenerated `pihole.toml` to docs for reference
  - For some reason a template doesn't exist anywhere upstream. [docs](https://docs.pi-hole.net/ftldns/configfile/) point you to just read the config file pihole autogenerates, which includes all the options and documentation. Unfortunately that's not useful when we're generating `pihole.toml` from module options

williamvds commented on nixos/modules/services/web-apps/pihole-web.nix:
```diff
@@ -0,0 +1,144 @@
+      description = "Temperature display unit";
+      default = "C";
+      example = "F";
+    };
```
Removed this with the v6 changes

williamvds commented on nixos/modules/services/web-apps/pihole-web.nix:
```diff
@@ -0,0 +1,144 @@
+        add_header X-Permitted-Cross-Domain-Policies "none";
+        add_header Referrer-Policy "same-origin";
+      '';
+    };
```
Removed now - pihole-ftl starts its own HTTP server, so setting up a reverse proxy is optional, and will be left as an exercise to the user

williamvds commented:
@purepani took a while, but I've got it back into ship shape with the v6 changes.
Would a appreciate a re-review when you have some time
Once happy I'll rebase into the usual commits.

guylamar2006 commented on nixos/modules/services/networking/pihole-ftl-setup-script.nix:
```diff
@@ -0,0 +1,70 @@
+  LoginAPI
+
+  ${builtins.concatStringsSep "\n" (
+    map (payload: lib.pipe payload [
+      lib.strings.escapeShellArg
+      (payload: "addList ${payload}")
+    ]) payloads
```
```suggestion
    map (
      payload:
      lib.pipe payload [
        lib.strings.escapeShellArg
        (payload: "addList ${payload}")
      ]
    ) payloads
```

guylamar2006 commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 361571`

Logs: https://github.com/guylamar2006/nixpkgs-review-gha/actions/runs/14560515607


---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>pihole</li>
    <li>pihole-ftl</li>
    <li>pihole-web</li>
  </ul>
</details>


---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>pihole</li>
    <li>pihole-ftl</li>
    <li>pihole-web</li>
  </ul>
</details>

purepani commented:
I'll try and take a look tomorrow. Feel free to ping me in case I forget :)

Pandapip1 commented on nixos/modules/services/networking/pihole-ftl.md:
```diff
@@ -0,0 +1,124 @@
+    queryLogDeleter.enable = true;
+    lists = [
+      {
+        url = "https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts";
```
Consider:

```suggestion
        url = "https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts";
        # Or: url = "file://${pkgs.stevenblack-blocklist}/hosts";
```

williamvds replied:
I disagree here - one of the nice things about Pihole is that it can update these adlists imperatively, to keep them up to date. e.g. StevenBlack's adlist gets updated ~every few days.

While you can possibly use NixOS auto-updates with nixpkgs-unstable to achieve something similar, providing a live URL is the standard way to use Pihole, so users will probably want to do the same.

williamvds replied:
Just noticed you suggested putting it in a comment. I'm not really against it, but I would prefer to add a warning that it won't be automatically updated

Pandapip1 replied:
This is fine. It's just that one of the nice things I like about NixOS is that generally your configurations stay static. So how about:

```suggestion
        url = "https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts";
        # Or, if you prefer updating your blocklist alongside the rest of your system:
        # url = "file://${pkgs.stevenblack-blocklist}/hosts";
        # This will, of course, not perform automatic updates
```

purepani replied:
If it's not so difficult to do, it would be good to at least have the option for a static configuration.

williamvds replied:
Added this to the example config

purepani reviewed (changes requested):
Just some small comments.
I'm not very familiar with resholve so I can't exactly review that part of the package definition, but otherwise the package definition looks fine.

The comments are not huge, but some of them would affect the api, so better to decide before merging

williamvds commented on nixos/modules/services/networking/pihole-ftl.nix:
```diff
@@ -0,0 +1,433 @@
+      '';
+    };
+
+    configDirectory = mkOption {
+      type = types.path;
+      default = "/etc/pihole";
+      internal = true;
+      description = ''
+        Path for pihole configuration.
+        pihole does not currently support any path other than /etc/pihole.
+      '';
+    };
```
Makes sense - I previously thought `internal` was the best method for that

williamvds replied:
You can still access internal options from other modules, so I think both internal and readOnly are appropriate.
In the future if I can make pihole no longer depend on `/etc/pihole`, we can probably remove the whole option

williamvds commented on nixos/modules/services/networking/pihole-ftl.nix:
```diff
@@ -0,0 +1,433 @@
+      '';
+    };
+
+    logDirectory = mkOption {
+      type = types.path;
+      default = "/var/log/pihole";
+      description = ''
+        Path for pihole log files.
+        pihole does not support any path other than /var/log/pihole.
+      '';
+    };
```
This path is now properly configurable in v6, so it can be set by users

purepani replied:
The description should probably be changed then.

williamvds commented on nixos/modules/services/networking/pihole-ftl.nix:
```diff
@@ -0,0 +1,433 @@
+
+    openFirewall = mkOption {
+      type = types.bool;
+      default = true;
```
I went looking through nixpkgs, it's a bit of a mixed bag as to whether these options default to true. Funnily dnamasq doesn't even have an `openFirewall` option.

Really, pihole is pretty useless without at least the DHCP/DNS server being exposed to the local network. But better to be explicit about these things, I guess.

I'll change the default to false, and update the example config.

purepani replied:
I can imagine you might not want to open the firewall if you're testing something for example, but yeah I do agree it's not really useful without it. To me, as you said, it's better to be explicit.

purepani replied:
Oh yeah, also if this is put behind a reverse proxy, there's no need to open the firewall.

williamvds replied:
No reverse proxy involved in the module, the pihole-FTL daemon acts as a webserver itself. So the option is to open the webserver's ports for external access. As mentioned before, I've left any reverse proxy setup as an exercise to the user.

purepani replied:
Yeah the edit is what I mean :smile:

williamvds commented on nixos/modules/services/web-apps/pihole-web.nix:
```diff
@@ -0,0 +1,107 @@
+        type = types.listOf (
+          types.oneOf [
+            types.port
+            types.str
```
The whole option submodule stuff? I thought providing a structured value was an improvement over the obscure suffix system they use, but I don't feel too strongly about it

purepani replied:
Oh wait, I just didn't read it correctly. For some reason I though you were parsing the strings, but it's a pihole thing.
Maybe put a link to the documentation for the suffix system stuff.

williamvds replied:
Alas, there's no documentation other than in the template `pihole.toml` I've included and mentioned in the manual. I summarised it in the option description, do you think that's sufficient?

purepani replied:
If that's the only documentation for it, then yeah that's fine.

williamvds commented:
Thanks, addressed the open comments, collapsed some threads

purepani reviewed (approved):
LGTM

williamvds commented:
Working on that this evening, thanks!

averyvigolo committed:
nixos/pihole-ftl: init

Add a module for pihole-ftl, which allows declaratively defining the
pihole.toml config file.

Also provide options for adlists to use, which can be added through the pihole
script (packaged as "pihole"). Other state such as clients and groups require
complex database operations, which is normally performed by the pihole
webapp (packaged as "pihole-web").

Extend the dnsmasq module to avoid duplication, since pihole-ftl is a soft-fork
of dnsmasq which maintains compatibility.

Provide the pihole script in `environment.systemPackages` so pihole-ftl can be
easily administrated.

averyvigolo committed:
nixos/pihole-web: init

Pihole's dashboard is a web app which visualises statistics from pihole-FTL
(i.e. dnsmasq), shows query logs, and allows configuration.

With this module, configuration is largely declarative and immutable, so
settings can't be changed, but they can be viewed from the webpage.

The admin page also allows regenerating the DNS ("gravity") database.

williamvds commented:
Squashed commits down and amended descriptions.
I'll leave any rebasing for the merger.

Did a last test of the modules with my own device's config and everything appears to be in working order.

purepani commented:
I just think someone with commit access needs to see it :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5448

jwiegley commented:
This looks good to me, I'll try installing it on my machine now and then I can merge it.

jwiegley commented on nixos/modules/services/web-apps/pihole-web.md:
```diff
@@ -0,0 +1,19 @@
+{
+  services.pihole-web = {
+    enable = true;
+    port = 80;
```
There is no `port` option for `pihole-web`. This example should be:
```
    pihole-web = {
      enable = true;
      ports = [
        "80r"
      ];
    };
```

williamvds replied:
Fixed, thanks

jwiegley reviewed (changes requested):
I received this error when trying to build your feature into my configuration:
```
error:
        failed to render manual manual.md

caused by:

Identifiers present in the source must have a mapping in the redirects file.
    - module-services-networking-pihole-ftl-administration
    - module-services-networking-pihole-ftl-configuration-inherit-dnsmasq
    - module-services-networking-pihole-ftl-configuration
    - module-services-web-apps-pihole-web-configuration
    - module-services-networking-pihole-ftl-configuration-multiple-interfaces
    - module-services-networking-pihole-ftl
    - module-services-web-apps-pihole-web
```

Have I incorrectly configured my flake?
```
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
    nixpkgs-pr.url = "github:williamvds/nixpkgs/add_pihole";
    nixos-hardware.url = "github:NixOS/nixos-hardware";

    nixos-logwatch = {
      url = "github:SFrijters/nixos-logwatch";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };

  outputs =
    { nixpkgs, nixpkgs-pr, nixos-hardware, nixos-logwatch, ... }:
    {
      formatter.x86_64-linux =
        nixpkgs.legacyPackages.x86_64-linux.nixfmt-rfc-style;

      nixosConfigurations.vulcan = nixpkgs.lib.nixosSystem {
        system = "x86_64-linux";
        modules = [
          nixos-logwatch.nixosModules.logwatch
          { nixpkgs.overlays = [
              (final: prev: {
                inherit (nixpkgs-pr.legacyPackages.${prev.system})
                  pihole pihole-ftl pihole-web;
              })
            ];
          }
          "${nixpkgs-pr}/nixos/modules/services/web-apps/pihole-web.nix"
          "${nixpkgs-pr}/nixos/modules/services/networking/pihole-ftl.nix"
          ./configuration.nix
          nixos-hardware.nixosModules.apple-t2
        ];
      };
    };
}
```

purepani commented:
I think that's the NixOS manual. I'm guessing that because you're only overriding the modules, it's not overriding the redirects file, so you can't build the manual.

purepani commented:
I think that's the NixOS manual. I'm guessing that because you're only overriding the modules, it's not overriding the redirects file, so you can't build the manual.

purepani commented:
Maybe using `pkgs.applyPatches` to apply the entire pr directly will work better.

jwiegley commented:
OK, I am rebuilding now and will come back once it's completed.

jwiegley commented:
@purepani I've reinstalled using the following flake:
```
{
  inputs = {
    # nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
    nixpkgs.url = "github:williamvds/nixpkgs/add_pihole";
    nixos-hardware.url = "github:NixOS/nixos-hardware";

    nixos-logwatch = {
      url = "github:SFrijters/nixos-logwatch";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };

  outputs =
    { nixpkgs, nixos-hardware, nixos-logwatch, ... }:
    let system = "x86_64-linux"; in {
      formatter.x86_64-linux =
        nixpkgs.legacyPackages.x86_64-linux.nixfmt-rfc-style;

      nixosConfigurations.vulcan = nixpkgs.lib.nixosSystem {
        inherit system;
        modules = [
          nixos-logwatch.nixosModules.logwatch
          ./configuration.nix
          nixos-hardware.nixosModules.apple-t2
        ];
      };
    };
}
```
However, although the service declarations worked out and I could see pihole building, none of the core services are present in systemd:
```
[root@vulcan:/etc/nixos]# systemctl list-units | grep -i hole
  podman-pihole.service                                                                                                                                     loaded active running   podman-pihole.service
  pihole-ftl-log-deleter.timer                                                                                                                              loaded active waiting   Pi-hole FTL DNS query log deleter
```
The podman-pihole is something else that I've been using, waiting for this PR.

Have I configured it incorrectly? Here is my excerpt from `configuration.nix`:
```
  services = rec {
    hardware.bolt.enable = true;

    # cockpit.enable = true;

    pihole-web = {
      enable = true;
      ports = [
        "9997r"
      ];
    };

    pihole-ftl = {
      enable = true;
      openFirewallDHCP = true;
      queryLogDeleter.enable = true;
      lists = [
        {
          url = "https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts";
          description = "Steven Black's unified adlist";
        }
      ];
      settings = {
        dns = {
          domainNeeded = true;
          expandHosts = true;
          interface = "enp4s0";
          listeningMode = "BIND";
          upstreams = [ "192.168.50.1#53" ];
        };
        dhcp = {
          active = false;
          # active = true;
          router = "192.168.50.1";
          start = "192.168.50.2";
          end = "192.168.50.255";
          leastTime = "1d";
          ipv6 = true;
          multiDNS = true;
          hosts = [
            # Static address for the current host
            "aa:bb:cc:dd:ee:ff,192.168.10.1,${config.networking.hostName},infinite"
          ];
          rapidCommit = true;
        };
        misc.dnsmasq_lines = [
          # This DHCP server is the only one on the network
          "dhcp-authoritative"
          # Source: https://data.iana.org/root-anchors/root-anchors.xml
          "trust-anchor=.,38696,8,2,683D2D0ACB8C9B712A1948B27F741219298D0A450D612C483AF444A4C0FB2B16"
        ];
      };
    };
  };
```

purepani commented:
I somewhat idiotically removed the code from when I tested it. Let me give it another shot.

williamvds commented:
Slowly struggling to reproduce your issue, trying out `applyPatches`, got it mostly evaluating, but now I need to deal with the incompatibilities with my config and nixpkgs-unstable

```nix
  mercury =
    let
      system = "x86_64-linux";
      base = inputs.nixpkgs-unstable;
      patched = ((import base { inherit system; }).applyPatches {
        name = "nixpkgs-patched";
        src = base;
        patches = [
          (builtins.fetchurl {
            url = "https://github.com/NixOS/nixpkgs/pull/361571.patch";
            sha256 = "sha256:11rab4n8j5i60i4i5fprkjjzn2har60dvfqagra5rq89yip4nd53";
          })
        ];
      });
    in
    base.lib.nixosSystem {
      inherit system;
      specialArgs = inputs;
      pkgs = import patched { inherit system; };
      baseModules = import "${patched}/nixos/modules/module-list.nix";
      modules = [
        ./mercury
      ];
    };
```

williamvds commented:
Fixed the gravity DB creation on the first run with a `kill -SIGRTMIN`.

Here's a VM NixOS config I've used for testing. Used with the above `nixosSystem` definition.

<details><summary>vm.nix</summary>

```nix
{
  config,
  lib,
  nixpkgs-patched,
  ...
}:

with lib;

{
  imports = [
    "${nixpkgs-patched}/nixos/modules/profiles/minimal.nix"
    "${nixpkgs-patched}/nixos/modules/profiles/qemu-guest.nix"
    "${nixpkgs-patched}/nixos/modules/virtualisation/qemu-vm.nix"
  ];

  fileSystems."/" = {
    device = "/dev/disk/by-label/nixos";
    fsType = "ext4";
    autoResize = true;
  };

  boot = {
    growPartition = true;
    kernelParams = [ "console=ttyS0" ];
    loader.grub.device = "/dev/vda";
    loader.timeout = 0;
  };

  virtualisation = {
    cores = 4;
    graphics = false;
    qemu.options = [ "-serial mon:stdio" ];
    forwardPorts = [
      {
        from = "host";
        host.port = 5757;
        guest.port = 57;
      }
      {
        from = "host";
        host.port = 8080;
        guest.port = 80;
      }
    ];
  };

  networking.hostName = "vm";
  users.extraUsers.root.password = "";
  services.getty.autologinUser = "root";

  services.openssh.settings.PermitRootLogin = mkForce "yes";

  services.pihole-ftl = {
    enable = true;
    openFirewallDHCP = true;
    openFirewallWebserver = true;
    queryLogDeleter.enable = true;
    lists = [
      {
        url = "https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts";
        # Alternatively, use the file from nixpkgs. Note its contents won't be
        # automatically updated by Pi-hole, as it would with an online URL.
        # url = "file://${pkgs.stevenblack-blocklist}/hosts";
        description = "Steven Black's unified adlist";
      }
    ];
    settings = {
      dns = {
        domainNeeded = true;
        expandHosts = true;
        interface = "eth0";
        listeningMode = "BIND";
        upstreams = [ "1.1.1.1" "1.0.0.1" ];
      };
      dhcp = {
        active = true;
        router = "192.168.10.1";
        start = "192.168.10.2";
        end = "192.168.10.254";
        leastTime = "1d";
        ipv6 = true;
        multiDNS = true;
        hosts = [
          # Static address for the current host
          "aa:bb:cc:dd:ee:ff,192.168.10.1,${config.networking.hostName},infinite"
        ];
        rapidCommit = true;
      };
      misc.dnsmasq_lines = [
        # This DHCP server is the only one on the network
        "dhcp-authoritative"
        # Source: https://data.iana.org/root-anchors/root-anchors.xml
        "trust-anchor=.,38696,8,2,683D2D0ACB8C9B712A1948B27F741219298D0A450D612C483AF444A4C0FB2B16"
      ];
    };
  };

  services.pihole-web = {
    enable = true;
    ports = [ 80 ];
  };
}
```
</details> 

```
$ nix build .#nixosConfigurations.vm.config.system.build.vm && ./result/bin/run-vm-vm
```

purepani commented:
I must have been thinking of something else that I tested when I said I tested it. The original one didn't seem to work. Apologies if that ended up confusing you at some point

williamvds commented:
Pushed another quick fix for formatting and sorting, and another typo in the pihole-FTL manual page (`leastTime` -> `leaseTime`)

@jwiegley your second config looks fine to me, I'm not sure why you'd only get the log deleter timer service. All the other services get enabled automatically with `pihole-ftl.enable`.
My main guess is that systemd services weren't reloaded properly when you activated system. Could poke around in `/etc/systemd/system/` to see if you find the expected `pihole-ftl` and `pihole-ftl-setup` services? Otherwise maybe inspect the NixOS config, eg. `nix eval .#nixosConfigurations.vulcan.config.systemd.services.pihole-ftl`

jwiegley commented:
@williamvds Thank you, I pulled, rebuilt, and I do see a systemd service for pihole FTL in `/etc/systemd/system`. However, I still have some difficulties loading. Here is the problem that `pihole-ftl-setup` reports during system rebuild:
```
systemd[1]: Starting Pi-hole FTL setup...
pihole-ftl-setup-start[2785159]: Please enter your password:
pihole-ftl-setup-start[2785188]: stty: 'standard input': Inappropriate ioctl for device
systemd[1]: pihole-ftl-setup.service: Main process exited, code=exited, status=1/FAILURE
systemd[1]: pihole-ftl-setup.service: Failed with result 'exit-code'.
systemd[1]: Failed to start Pi-hole FTL setup.
```

How should I configure things so it doesn't ask for a password in a background service?

Then pihole-ftl also complained about a missing command:
```
pihole-FTL[...]: INFO: ########## FTL started on vulcan! ##########
pihole-FTL[...]: INFO: FTL branch: master
pihole-FTL[...]: INFO: FTL version: v6.1
pihole-FTL[...]: INFO: FTL commit: builtfromreleasetarball
pihole-FTL[...]: INFO: FTL date: 1970-01-01
pihole-FTL[...]: INFO: FTL user: pihole
pihole-FTL[...]: INFO: Compiled for x86_64 (compiled locally) using gcc (GCC) 14.2.1 20250322
pihole-FTL[...]: INFO: Parsed config file /etc/pihole/pihole.toml successfully
pihole-FTL[...]: WARNING: Unable to read PID from file: No such file or directory
pihole-FTL[...]: INFO: PID file does not exist or not readable
pihole-FTL[...]: INFO: No other running FTL process found.
pihole-FTL[...]: WARNING: Unable to write PID to file: Permission denied
pihole-FTL[...]: INFO: PID of FTL process: 2783237
pihole-FTL[...]: sh: line 1: ip: command not found
pihole-FTL[...]: sh: line 1: ip: command not found
```

Note that for testing I'm running pihole DNS at port 5353, and I indeed I see it there and can query it just fine:
```
[root@vulcan:/etc/nixos]# netstat -nl
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State
tcp        0      0 192.168.50.182:5353     0.0.0.0:*               LISTEN
tcp        0      0 127.0.0.1:5353          0.0.0.0:*               LISTEN
```

However, the web service is no where to be seen. I'm trying to enable it using:
```
    pihole-web = {
      enable = true;
      ports = [
        "9997r"
      ];
    };
```

Once I can resolve all of these issues, this is looking good to go in!

williamvds commented:
Glad you've gotten a bit further - to be honest I've no idea why the setup script is asking for a password. There shouldn't be a `sudo` in there, it should all just work since it runs as the pihole user.

Could you try getting a trace by adding something equivalent to this?
```nix
  systemd.services.pihole-ftl-setup = {
    script = lib.mkForce ''
      set -x
      ${import "${nixpkgs-patched}/nixos/modules/services/networking/pihole-ftl-setup-script.nix" {
          inherit
            config
            lib
            pkgs
            ;

          cfg = config.services.pihole-ftl;
      }}
    '';
  };
```

I'll look into the `ip` command missing thing, I'm also not sure why or where FTL is trying to run that.

And finally, the pihole-web service no longer exists. In v6, the FTL process itself is the webserver.

jwiegley commented:
Hmm.. your debug expression yields "Cannot coerce a set to a string" when trying to interpolate the results of calling that function from the setup nix file...

williamvds commented:
I'm guessing that in the `import "${nixpkgs-patched}/nixos/modules..."` bit, you've replaced `nixpkgs-patched` with `pkgs`? It needs to be replaced with a flake path from your inputs instead, not the imported flake.

jwiegley commented:
@williamvds Ah, I see, that makes sense.

Since the `pihole-web` service is not being used anymore, should those files and their documentation be removed from the PR? The service still evaluates presently.

jwiegley commented:
@williamvds Here is the debug output for `pihole-ftl-setup`:
```
+ set -eo pipefail
+ pihole=/nix/store/3cdrz34dwqp4rxhwm62q6ia9ixniaz9d-pihole-6.0.6/bin/pihole
+ jq=/nix/store/2yi7k5ay6xj73rbfqf170gms922rjm2d-jq-1.7.1-bin/bin/jq
+ '[' '!' -f /var/lib/pihole/gravity.db ']'
+ source /nix/store/3cdrz34dwqp4rxhwm62q6ia9ixniaz9d-pihole-6.0.6/usr/share/pihole/advanced/Scripts/api.sh
+ source /nix/store/3cdrz34dwqp4rxhwm62q6ia9ixniaz9d-pihole-6.0.6/usr/share/pihole/advanced/Scripts/utils.sh
+ any_failed=0
+ for i in 1 2 3
+ TestAPIAvailability
+ local chaos_api_list authResponse authStatus authData
++ /nix/store/vamfsrs9sqhibjcxdqzivmhwhlw0chc8-bind-9.20.7-dnsutils/bin/dig +short chaos txt local.api.ftl @127.0.0.1
+ chaos_api_list='"http://127.0.0.1:8082/api/" "https://127.0.0.1:8444/api/" "http://[::1]:8082/api/" "https://[::1]:8444/api/"'
+ '[' -z '"http://127.0.0.1:8082/api/" "https://127.0.0.1:8444/api/" "http://[::1]:8082/api/" "https://[::1]:8444/api/"' ']'
+ '[' '"http://127.0.0.1:8082/api/" "https://127.0.0.1:8444/api/" "http://[::1]:8082/api/" "https://[::1]:8444/api/"' '!=' '"http://127.0.0.1:8082/api/" "https://127.0.0.1:8444/api/" "http://[::1]:8082/api/" "https://[::1]:8444/api/"' ']'
+ '[' -n '"http://127.0.0.1:8082/api/" "https://127.0.0.1:8444/api/" "http://[::1]:8082/api/" "https://[::1]:8444/api/"' ']'
+ API_URL='"http://127.0.0.1:8082/api/"'
+ API_URL='"http://127.0.0.1:8082/api/'
+ API_URL=http://127.0.0.1:8082/api/
++ /nix/store/lr1pnwv3l0ks6w5fbl7yn80lskgaksi6-curl-8.12.1-bin/bin/curl --connect-timeout 2 -skS -w '%{http_code}' http://127.0.0.1:8082/api/auth
+ authResponse='{"session":{"valid":false,"totp":false,"sid":null,"validity":-1,"message":"no SID provided"},"took":3.9815902709960938e-05}401'
++ printf %s '{"session":{"valid":false,"totp":false,"sid":null,"validity":-1,"message":"no SID provided"},"took":3.9815902709960938e-05}401'
++ /nix/store/cg09nslw3w6afyynjw484b86d47ic1cb-coreutils-9.7/bin/tail -c 3
+ authStatus=401
++ printf %s '{"session":{"valid":false,"totp":false,"sid":null,"validity":-1,"message":"no SID provided"},"took":3.9815902709960938e-05}'
+ authData='{"session":{"valid":false,"totp":false,"sid":null,"validity":-1,"message":"no SID provided"},"took":3.9815902709960938e-05}'
+ '[' '!' 401 = 200 ']'
+ '[' '!' 401 = 401 ']'
+ '[' 401 = 200 ']'
++ echo '{"session":{"valid":false,"totp":false,"sid":null,"validity":-1,"message":"no SID provided"},"took":3.9815902709960938e-05}'
++ /nix/store/2yi7k5ay6xj73rbfqf170gms922rjm2d-jq-1.7.1-bin/bin/jq --raw-output .session.totp
+ needTOTP=false
+ break
+ '[' -n '' ']'
+ break
+ LoginAPI
+ '[' -z '' ']'
+ TestAPIAvailability
+ local chaos_api_list authResponse authStatus authData
++ /nix/store/vamfsrs9sqhibjcxdqzivmhwhlw0chc8-bind-9.20.7-dnsutils/bin/dig +short chaos txt local.api.ftl @127.0.0.1
+ chaos_api_list='"http://127.0.0.1:8082/api/" "https://127.0.0.1:8444/api/" "http://[::1]:8082/api/" "https://[::1]:8444/api/"'
+ '[' -z '"http://127.0.0.1:8082/api/" "https://127.0.0.1:8444/api/" "http://[::1]:8082/api/" "https://[::1]:8444/api/"' ']'
+ '[' '"http://127.0.0.1:8082/api/" "https://127.0.0.1:8444/api/" "http://[::1]:8082/api/" "https://[::1]:8444/api/"' '!=' '"http://127.0.0.1:8082/api/" "https://127.0.0.1:8444/api/" "http://[::1]:8082/api/" "https://[::1]:8444/api/"' ']'
+ '[' -n '"http://127.0.0.1:8082/api/" "https://127.0.0.1:8444/api/" "http://[::1]:8082/api/" "https://[::1]:8444/api/"' ']'
+ API_URL='"http://127.0.0.1:8082/api/"'
+ API_URL='"http://127.0.0.1:8082/api/'
+ API_URL=http://127.0.0.1:8082/api/
++ /nix/store/lr1pnwv3l0ks6w5fbl7yn80lskgaksi6-curl-8.12.1-bin/bin/curl --connect-timeout 2 -skS -w '%{http_code}' http://127.0.0.1:8082/api/auth
+ authResponse='{"session":{"valid":false,"totp":false,"sid":null,"validity":-1,"message":"no SID provided"},"took":4.482269287109375e-05}401'
++ printf %s '{"session":{"valid":false,"totp":false,"sid":null,"validity":-1,"message":"no SID provided"},"took":4.482269287109375e-05}401'
++ /nix/store/cg09nslw3w6afyynjw484b86d47ic1cb-coreutils-9.7/bin/tail -c 3
+ authStatus=401
++ printf %s '{"session":{"valid":false,"totp":false,"sid":null,"validity":-1,"message":"no SID provided"},"took":4.482269287109375e-05}'
+ authData='{"session":{"valid":false,"totp":false,"sid":null,"validity":-1,"message":"no SID provided"},"took":4.482269287109375e-05}'
+ '[' '!' 401 = 200 ']'
+ '[' '!' 401 = 401 ']'
+ '[' 401 = 200 ']'
++ echo '{"session":{"valid":false,"totp":false,"sid":null,"validity":-1,"message":"no SID provided"},"took":4.482269287109375e-05}'
++ /nix/store/2yi7k5ay6xj73rbfqf170gms922rjm2d-jq-1.7.1-bin/bin/jq --raw-output .session.totp
+ needTOTP=false
+ break
+ '[' -n '' ']'
+ '[' '' = false ']'
+ '[' -r /etc/pihole/cli_pw ']'
+ '[' '' = verbose ']'
+ '[' -z '' ']'
+ echo 'Please enter your password:'
Please enter your password:
+ secretRead
++ /nix/store/cg09nslw3w6afyynjw484b86d47ic1cb-coreutils-9.7/bin/stty -g
stty: 'standard input': Inappropriate ioctl for device
+ stty_orig=
```

williamvds commented:
Thanks for the logs, I suspected you had some API authentication enabled. I'll test that out, there must be a way for the script to automatically authenticate itself.

For pihole-web, there's no systemd service, but they remain as separate NixOS modules. I can merge the web module into FTL, move the options under `pihole-ftl.web` perhaps. Do you think that's more reasonable / simpler?

jwiegley commented:
@williamvds Oh, if pihole-web has a purpose then the current arrangement is fine. Looking forward to trying your next update re: the setup!

williamvds commented:
Okay, I've reproduced your issue by setting `webserver.api.pwhash` to a random value (for some reason `pihole setpassword` did nothing, even after setting `misc.readOnly=false`), and `webserver.api.cli_pw=false`. It's clear you need to enable the autogenerated CLI password if you want to be able to add lists automatically like this.

I'll add an assert to check that `cli_pw=true` if your config has defined any lists.

As a sidenote, this CLI authentication business is pretty convoluted. It would be neater if pihole-FTL created some socket file owned by the pihole user, and the pihole script sent requests there instead. It would be safe to assume only someone with sufficient permissions could reach that socket and therefore further authentication is unnecessary.

williamvds commented:
And the setpassword thing is because it runs the `pihole-FTL` executable to set the config option, and that executable complains unless you give it all the capabilities it needs to run the full daemon.

They run the executable because for some reason they don't let sessions authenticated by the CLI password modify config. Despite the fact you need to read private state files to get that password. :shrug:

It's looking like a pain to give the pihole-FTL executable all the capabilities just to get the `pihole setpassword` command working. Which won't work anyway unless you set `misc.readOnly=false` which is kind of incompatible with the NixOS module anyway. I'm willing to say that's unsupported.

Finally, onto the `ip` command

jwiegley commented:
@williamvds Hi Avery, it still errors for me: https://gist.github.com/jwiegley/5c376a307e0054bc81e2ca2fd7738886

Is there anything else I need to change?

williamvds commented:
> @williamvds Hi Avery, it still errors for me: https://gist.github.com/jwiegley/5c376a307e0054bc81e2ca2fd7738886
> 
> Is there anything else I need to change?

Hey @jwiegley, could you share your config? With the latest updates, you should hit an assertion error if `settings.webserver.api.cli_pw` isn't set to true, which is what fixes this authentication issue

jwiegley commented:
@williamvds Here is my config: https://gist.github.com/jwiegley/c2e3237dc37110afb89edb564903a54e

purepani commented:
Does it have maybe to do with the pihole oci container at the bottom?

jwiegley commented:
I'd like to run the pihole service at different ports, so that I can have both running until I'm ready to switch over...

jwiegley commented:
Well, disabling it will bring down Internet for everyone in my house, so it'll have to wait until there's a window when no one needs it.

williamvds commented:
@jwiegley could you try setting `service.pihole-ftl.settings.webserver.api.cli_pw = true` in your config? I don't see it in there

jwiegley commented:
@williamvds I now have the following:
```
  services = rec {
    pihole-ftl = {
      enable = true;
      settings = {
        webserver = {
          port = 7272;
          api.cli_pw = true;
        };
```
but there is no change, it's still asking for my password as before.

williamvds commented:
Would you be able to share your (redacted) `/etc/pihole/pihole.toml`? Just as a sanity check to see that cli_pw is set there

jwiegley commented:
Yes, it seems to be set there:
```
[webserver]
port = 7272
tls = "/var/lib/pihole/tls.pem"
[webserver.api]
cli_pw = true
```

jwiegley commented:
Btw, I am at matrix.org in several of the Nix channels, if you'd like to do this less async.

evilbulgarian commented:
Thanks @williamvds for this great work! Not sure if its possible with pihole, but since this is suppose to be a replacement for dnsmasq, would it be possible to have the nixos configuration be similar? Currently I use dnsmasq to configure my local network across multiple vlans and other more advanced options exposed. Does pihole have a minimal set of addons to dnsmasq.conf if so we can maybe merge how dnsmasq nixos module configures dhcp / dns and add the pihole bits that way it would be easier for people to migrate from dnsmasq setup to pihole. Just a thought. Thanks again.

williamvds commented:
Hey @evilbulgarian, so I added the option `services.pihole-ftl.useDnsmasqConfig` which essentially is used for that.

It'll merge `services.dnsmasq.settings` into your pihole options. It's not perfect, because pihole generates some dnsmasq config of its own, causing it to fall over and complain about duplicate options.

So you might have to disable a line or two in your dnsmasq settings to make it happy.

The pihole GUI probably won't be fully accurate because it pulls from pihole.toml, but everything should still work just fine.

We could probably spend some time getting the imported dnsmasq settings to play nice with pihole's generated config, but that could be left as a later improvement.

evilbulgarian commented:
> Hey @evilbulgarian, so I added the option `services.pihole-ftl.useDnsmasqConfig` which is essentially for the purpose of migrating from plain dnsmasq.
> 
> It'll merge `services.dnsmasq.settings` into your pihole options. It's not perfect, because pihole generates some dnsmasq config of its own, which may cause it to fall over and complain about duplicate options.
> 
> So you might have to disable a line or two in your dnsmasq settings to make it happy.
> 
> The pihole GUI probably won't be fully accurate because it pulls from pihole.toml, but everything should still work just fine.
> 
> We could probably spend some time getting the imported dnsmasq settings to play nice with pihole's generated config, but that could be left as a later improvement.

nice great to see you already thought of it! will test it out.

ashuramaruzxc commented:
I don't see anyone to test this on aarch64 so i will test it out on my rpi5 with nixos, to check whether if it works as intended

evilbulgarian commented:
> > Hey @evilbulgarian, so I added the option `services.pihole-ftl.useDnsmasqConfig` which is essentially for the purpose of migrating from plain dnsmasq.
> > It'll merge `services.dnsmasq.settings` into your pihole options. It's not perfect, because pihole generates some dnsmasq config of its own, which may cause it to fall over and complain about duplicate options.
> > So you might have to disable a line or two in your dnsmasq settings to make it happy.
> > The pihole GUI probably won't be fully accurate because it pulls from pihole.toml, but everything should still work just fine.
> > We could probably spend some time getting the imported dnsmasq settings to play nice with pihole's generated config, but that could be left as a later improvement.
> 
> nice great to see you already thought of it! will test it out.

```

       error: undefined variable 'dnsmasqConf'
       at /nix/store/plqfl5r05rihwsx3gd0dibydkypbwjys-source/nixos/modules/services/networking/pihole-ftl.nix:246:39:
          245|       (lib.optionalAttrs cfg.useDnsmasqConfig {
          246|         misc.dnsmasq_lines = lib.pipe dnsmasqConf [
             |                                       ^
          247|           builtins.readFile
```
error I am getting once I set the flag you mentioned @williamvds. I will try and debug and fix and will submit PR if I figure it out.

evilbulgarian commented:
> > > > Hey @evilbulgarian, so I added the option `services.pihole-ftl.useDnsmasqConfig` which is essentially for the purpose of migrating from plain dnsmasq.
> > > > It'll merge `services.dnsmasq.settings` into your pihole options. It's not perfect, because pihole generates some dnsmasq config of its own, which may cause it to fall over and complain about duplicate options.
> > > > So you might have to disable a line or two in your dnsmasq settings to make it happy.
> > > > The pihole GUI probably won't be fully accurate because it pulls from pihole.toml, but everything should still work just fine.
> > > > We could probably spend some time getting the imported dnsmasq settings to play nice with pihole's generated config, but that could be left as a later improvement.
> > > 
> > > 
> > > nice great to see you already thought of it! will test it out.
> > 
> > 
> > ```
> > 
> >        error: undefined variable 'dnsmasqConf'
> >        at /nix/store/plqfl5r05rihwsx3gd0dibydkypbwjys-source/nixos/modules/services/networking/pihole-ftl.nix:246:39:
> >           245|       (lib.optionalAttrs cfg.useDnsmasqConfig {
> >           246|         misc.dnsmasq_lines = lib.pipe dnsmasqConf [
> >              |                                       ^
> >           247|           builtins.readFile
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > error I am getting once I set the flag you mentioned @williamvds. I will try and debug and fix and will submit PR if I figure it out.
> 
> Sorry, that was my mistake. A local bodge from when I was maintaining the package in a separate repo snuck in to the commit. I've pushed a fix

no problemo, now same error but unable to find `configFile`, here is the config I am using: ```
  services.pihole-web = {
    enable = true;
  };
  services.pihole-ftl = {
    enable = true;
    useDnsmasqConfig = true;
  };
  services.dnsmasq = {
    enable = false;
    settings = {
      server = [
        "127.0.0.1#5453"
...
```

williamvds commented:
@evilbulgarian 
If the option `config.services.dnsmasq.configFile`, that must mean you're somehow not using all the changes in this PR? 
I added `configFile` to the dnsmasq module in this commit https://github.com/NixOS/nixpkgs/pull/361571/commits/c876952e0429e90ebefafd0a318c381ff48f68e5

evilbulgarian commented:
> @evilbulgarian If the option `config.services.dnsmasq.configFile`, that must mean you're somehow not using all the changes in this PR? I added `configFile` to the dnsmasq module in this commit [c876952](https://github.com/NixOS/nixpkgs/commit/c876952e0429e90ebefafd0a318c381ff48f68e5)

Roger thanks, will change how my flake pulls in the modules.

williamvds commented:
@jwiegley upstream has fixed the DNS port issue impressively quickly, so I've updated the pi-hole package to 6.1. That should fix your test case.

jwiegley commented:
> @jwiegley upstream has fixed the DNS port issue impressively quickly, so I've updated the pi-hole package to 6.1. That should fix your test case.

It indeed fixes the error I was seeing, but things have now moved to a new error. This is after restarting pihole-ftl.service, which reports that it is running:
```
Communication error. Is FTL running?
+ exit 1
+ echo 'Retrying API shortly...'
Retrying API shortly...
+ /nix/store/cg09nslw3w6afyynjw484b86d47ic1cb-coreutils-9.7/bin/sleep .5s
+ LoginAPI
+ '[' -z '' ']'
+ TestAPIAvailability
+ local chaos_api_list authResponse authStatus authData apiAvailable DNSport utilsfile
+ readonly utilsfile=/nix/store/kzzsyyvsnbnlrarkz0fc6z3456apdz53-pihole-6.1/usr/share/pihole/advanced/Scripts/utils.sh
+ utilsfile=/nix/store/kzzsyyvsnbnlrarkz0fc6z3456apdz53-pihole-6.1/usr/share/pihole/advanced/Scripts/utils.sh
+ . /nix/store/kzzsyyvsnbnlrarkz0fc6z3456apdz53-pihole-6.1/usr/share/pihole/advanced/Scripts/utils.sh
++ getFTLConfigValue dns.port
++ /nix/store/r7059n9gjyacwk02pmkkx2wlcpr8rg3l-pihole-ftl-6.1/bin/pihole-FTL --config -q dns.port
+ DNSport=5353
++ /nix/store/vamfsrs9sqhibjcxdqzivmhwhlw0chc8-bind-9.20.7-dnsutils/bin/dig +short -p 5353 chaos txt local.api.ftl @127.0.0.1
+ chaos_api_list=';; Warning: Message parser reports malformed message packet.'
+ '[' -z ';; Warning: Message parser reports malformed message packet.' ']'
+ '[' ' Warning: Message parser reports malformed message packet.' '!=' ';; Warning: Message parser reports malformed message packet.' ']'
+ echo 'Communication error. Is FTL running?'
Communication error. Is FTL running?
+ exit 1
May 31 08:43:44 vulcan systemd[1]: pihole-ftl-setup.service: Main process exited, code=exited, status=1/FAILURE
░░ Subject: Unit process exited
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ An ExecStart= process belonging to unit pihole-ftl-setup.service has exited.
```

williamvds commented:
Okay, looks like the API requires the webserver to be enabled. Will add that to the assert

williamvds commented:
Now with your config you'll get this assertion error:
```
       error:
       Failed assertions:
       - The Pi-hole webserver must be enabled for lists set in services.pihole-ftl.lists to be automatically loaded on startup via the web API.
       services.pihole-ftl.settings.port must be defined, e.g. by enabling services.pihole-web.enable and defining services.pihole-web.port.
```

Rebased and resolved merge conflicts, moving the release notes to 25.11 (:cry:).
Hopefully the rebase fixes CI. I assume it was causing evaluation errors.

jwiegley commented:
@williamvds Some success! `pihole-ftl-setup.service` was able to restart. This evening when I can bring down my home network, I Will switch over to your pihole service and give it some final testing. Hopefully I can approve soon.

drupol commented on pkgs/by-name/pi/pihole-web/package.nix:
```diff
@@ -0,0 +1,51 @@
+  src = fetchFromGitHub {
+    owner = "pi-hole";
+    repo = "web";
+    rev = "v${finalAttrs.version}";
```
```suggestion
    tag = "v${finalAttrs.version}";
```

drupol commented on pkgs/by-name/pi/pihole-ftl/package.nix:
```diff
@@ -0,0 +1,86 @@
+  src = fetchFromGitHub {
+    owner = "pi-hole";
+    repo = "FTL";
+    rev = "v${finalAttrs.version}";
```
```suggestion
    tag = "v${finalAttrs.version}";
```

drupol commented on pkgs/by-name/pi/pihole-ftl/package.nix:
```diff
@@ -0,0 +1,86 @@
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin
+    cp pihole-FTL $out/bin
```
Replace this with a single line with `install`

drupol commented on pkgs/by-name/pi/pihole-ftl/package.nix:
```diff
@@ -0,0 +1,86 @@
+
+  passthru.settingsTemplate = ./pihole.toml;
+
+  meta = with lib; {
```
```suggestion
  meta = {
```

drupol commented on pkgs/by-name/pi/pihole-ftl/package.nix:
```diff
@@ -0,0 +1,86 @@
+  meta = with lib; {
+    description = "Pi-hole FTL engine";
+    homepage = "https://github.com/pi-hole/FTL";
+    license = licenses.eupl12;
+    maintainers = with maintainers; [ williamvds ];
+    platforms = platforms.linux;
```
```suggestion
    license = lib.licenses.eupl12;
    maintainers = with lib.maintainers; [ williamvds ];
    platforms = lib.platforms.linux;
```

drupol commented on nixos/modules/services/web-apps/pihole-web.nix:
```diff
@@ -0,0 +1,106 @@
+  ...
+}:
+
+with { inherit (lib) mkOption types; };
+
```
Is this really necessary?

drupol commented on nixos/modules/services/web-apps/pihole-web.nix:
```diff
@@ -0,0 +1,106 @@
+{
```
Perhaps we could have an option like: `openFirewall`? Many services are already using that.

drupol commented on pkgs/by-name/pi/pihole/package.nix:
```diff
@@ -0,0 +1,259 @@
+  ...
+}:
+
+(resholve.mkDerivation rec {
```
What is `resholve.mkDerivation`? Can we get rid of `rec` and use the `finalAttrs` pattern?

drupol commented on pkgs/by-name/pi/pihole/package.nix:
```diff
@@ -0,0 +1,259 @@
+  src = fetchFromGitHub {
+    owner = "pi-hole";
+    repo = "pi-hole";
+    rev = "v${version}";
```
```suggestion
    tag = "v${version}";
```

drupol commented on pkgs/by-name/pi/pihole/package.nix:
```diff
@@ -0,0 +1,259 @@
+    hash = "sha256-aEnv8Lhb5vf0yDyuriVTaUY1wcdVmTdqoK+KDHvT/Lw=";
+  };
+
+  patches = [
+    ./0001-Remove-sudo.patch
+    ./0002-Remove-unsupported-commands.patch
+    ./0003-Fix-redefinition-of-readonly-variable-utilsfile.patch
+  ];
```
I see that the filenames are pretty self-explanatory already, but it would be a good idea to add a comment explaining why those patches are needed and if they need to be removed at the next bump.

drupol commented on pkgs/by-name/pi/pihole/package.nix:
```diff
@@ -0,0 +1,259 @@
+{
+  lib,
+  fetchFromGitHub,
+  makeWrapper,
```
```suggestion
  makeBinaryWrapper,
```

Let's use the preferred way, as recommended in the nix manual.

drupol commented on pkgs/by-name/pi/pihole/package.nix:
```diff
@@ -0,0 +1,259 @@
+    # The installation script is sourced by advanced/Scripts/webpage.sh
+    cp --parents -r -t $scriptsDir/ 'automated install/' advanced/{Scripts,Templates}/
+
+    mkdir -p $out/bin
+
+    cp -t $out/bin pihole
```
Prefer a single line with `install`

drupol reviewed (commented):
Massive work, thanks!

I added a few comments.

jwiegley reviewed (approved):
This all looks good to me, jsut waiting on remaining comments by others to be resolved.

K900 commented:
Why is there a `sync.sh` in nixpkgs root now?

drupol commented:
> Why is there a `sync.sh` in nixpkgs root now?

https://github.com/NixOS/nixpkgs/pull/413974

dotlambda commented on nixos/modules/services/networking/pihole-ftl-setup-script.nix:
```diff
@@ -0,0 +1,82 @@
+  if [ ! -f '${cfg.stateDirectory}'/gravity.db ]; then
+    $pihole -g
+    # Send SIGRTMIN to FTL, which makes it reload the database, opening the newly created one
+    ${pkgs.procps}/bin/kill -s SIGRTMIN $(systemctl show --property MainPID --value ${config.systemd.services.pihole-ftl.name})
```
```suggestion
    ${lib.getExe' pkgs.procps "kill"} -s SIGRTMIN $(systemctl show --property MainPID --value ${config.systemd.services.pihole-ftl.name})

dotlambda commented on nixos/modules/services/networking/dnsmasq.nix:
```diff
@@ -172,7 +178,7 @@ in
         Type = "dbus";
         BusName = "uk.org.thekelleys.dnsmasq";
-        ExecStart = "${dnsmasq}/bin/dnsmasq -k --enable-dbus --user=dnsmasq -C ${dnsmasqConf}";
+        ExecStart = "${dnsmasq}/bin/dnsmasq -k --enable-dbus --user=dnsmasq -C ${cfg.configFile}";
```
```suggestion
        ExecStart = "${lib.getExe dnsmasq} -k --enable-dbus --user=dnsmasq -C ${cfg.configFile}";

dotlambda commented on nixos/modules/services/networking/pihole-ftl-setup-script.nix:
```diff
@@ -0,0 +1,82 @@
+    ${pkgs.procps}/bin/kill -s SIGRTMIN $(systemctl show --property MainPID --value ${config.systemd.services.pihole-ftl.name})
+  fi
+
+  source ${pihole}/usr/share/pihole/advanced/Scripts/api.sh
```
```suggestion
  source ${pihole}/share/pihole/advanced/Scripts/api.sh
```
and fix the package accordingly

dotlambda commented on pkgs/by-name/pi/pihole-ftl/package.nix:
```diff
@@ -0,0 +1,85 @@
+      --replace-fail "execv" "execvp"
+
+    substituteInPlace src/database/network-table.c \
+      --replace-fail "ip neigh show" "${iproute2}/bin/ip neigh show" \
```
```suggestion
      --replace-fail "ip neigh show" "${lib.getExe' iproute2 "ip"} neigh show" \

dotlambda commented on pkgs/by-name/pi/pihole-web/package.nix:
```diff
@@ -0,0 +1,51 @@
+    runHook postInstall
+  '';
+
+  meta = {
```
`homepage` and `changelog` missing

dotlambda commented on pkgs/by-name/pi/pihole/package.nix:
```diff
@@ -0,0 +1,257 @@
+  sqlite,
+  systemd,
+  util-linux,
+  stateDir ? "/etc/pihole",
```
```suggestion
  stateDir ? "/var/lib/pihole",

dotlambda commented on pkgs/by-name/pi/pihole/package.nix:
```diff
@@ -0,0 +1,257 @@
+  installPhase = ''
+    runHook preInstall
+
+    readonly scriptsDir=$out/usr/share/pihole
```
```suggestion
    readonly scriptsDir=$out/share/pihole

dotlambda commented on pkgs/by-name/pi/pihole/package.nix:
```diff
@@ -0,0 +1,257 @@
+      ];
+    };
+
+  meta = {
```
`homepage` and `changelog` missing

dotlambda commented on pkgs/by-name/pi/pihole/package.nix:
```diff
@@ -0,0 +1,257 @@
+    };
+
+  meta = {
+    description = "A black hole for Internet advertisements";
```
```suggestion
    description = "Black hole for Internet advertisements";
```
https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#meta-attributes

dotlambda commented on pkgs/by-name/pi/pihole/package.nix:
```diff
@@ -0,0 +1,257 @@
+  };
+
+  passthru = {
+    stateDir = stateDir;
```
```suggestion
    inherit stateDir;

drupol commented:
Hey don't be sorry, :) we are all humans and we make mistakes ! Thanks for doing a follow-up PR, so sad those comments, like most of the time, too late.

Al2Klimov commented:
That much for an adblocker? Hold my beer. :)

## With own resolver

```nix
{ ... }:
{
  services.bind = {
    enable = true;
    listenOn = [ ];
    listenOnIpv6 = [ ];
    extraOptions = "listen-on port 553 { any; };";
  };

  services.dnsmasq = {
    enable = true;
    resolveLocalQueries = false;
    settings = {
      no-hosts = true;
      no-resolv = true;
      no-poll = true;
      server = [ "127.0.0.1#553" ];

      addn-hosts = (fetchGit {
        url = "https://github.com/StevenBlack/hosts.git";
      }) + "/hosts";
    };
  };

  networking.firewall = {
    allowedTCPPorts = [ 53 ];
    allowedUDPPorts = [ 53 ];
  };
}
```

## With external resolver

```nix
{ ... }:
{
  services.dnsmasq = {
    enable = true;
    resolveLocalQueries = false;
    settings = {
      no-hosts = true;
      no-resolv = true;
      no-poll = true;
      server = [ "9.9.9.9" ];

      addn-hosts = (fetchGit {
        url = "https://github.com/StevenBlack/hosts.git";
      }) + "/hosts";
    };
  };

  networking.firewall = {
    allowedTCPPorts = [ 53 ];
    allowedUDPPorts = [ 53 ];
  };
}
```
